src/SourceLocator/Type/AutoloadSourceLocator.php

```diff
@@ -183,14 +183,16 @@ private function locateClassByName(string $className) : ?string
             stream_wrapper_register($protocol, self::class);
         }
 
-        class_exists($className);
+        try {
+            class_exists($className);
+        } finally {
```

Comment:
This is needed when the autoloader throws an exception which would ideally have also a test case.

Also, I later found out that autoloaders that use a `require` instead of `include` are problematic and still end in fatal error. It helped:

1) to use `restore_error_handler()` instead of `set_error_handler($previousErrorHandler)` although I'm not sure why.
2) instead of `return true` in the error handler I'm throwing and catching an exception immediately: https://github.com/phpstan/phpstan-src/commit/83bbc8122a7286b86f01288fa85957f34378814c#diff-b84fb81e70db3cdb6b4495c80b475a44 Another advantage of the exception is that only the first of the registered autoloaders is hit and not all of them. The `require` still throws a warning on older PHP versions but it's fixed in 7.4.2 (https://3v4l.org/CkDDJ). I didn't find a better solution.

Reply:
(This is what happens with the current code and unsuccessful require: https://3v4l.org/Nt7Ya)

Reply:
> use `restore_error_handler()` instead of `set_error_handler($previousErrorHandler)`

Yeah that seems a bit of an oversight, looks like a reasonable fix

> instead of `return true` in the error handler I'm throwing and catching an exception immediately

wow :vomiting_face: but still, it's all a big smashing hack, so understandable. Let's go with the throwing exception mechanism to solve those edge cases :+1:

Reply:
> use restore_error_handler() instead of set_error_handler($previousErrorHandler)

We cannot use `restore_error_handler()` because of `AutoloadSourceLocator::url_stat()`

> instead of return true in the error handler I'm throwing and catching an exception immediately

@ondrejmirtes See "WIP" commit please. Tests are failing. Did I missed something?

Reply:
`restore_error_handler()` added. See comment below.